### ISSUE TITLE
Tracing mode for unbacked SymInts using real data

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#112749]
### 🐛 Describe the bug

@aakhundov and I were talking about guard on data-dependent unbacked symint bugs, and it emerged that there might be a useful alternate tracing mode using real example inputs that might materially improve UX of unbacked symints. Here's what it looks like.

Suppose that you have a program that performs a conditional on unbacked SymInt:

```
x = y.nonzero()
if x.size(0) != 20:
     assert False
do_something()
```

This program will fail to trace in Dynamo, because Dynamo does not have a hint for `x.size(0)` and has no idea which way the conditional will go. However, it turns out that in many cases, there is some implicit invariants that mean the conditionals will always go one way or another--in the example above, we know that `x.size(0) == 20`, because it would error otherwise (if you write `torch._check(x.size(0) == 20)`, we can in fact automatically infer this, but in many cases the conditional is nontrivial to identify). What we would like to do is *infer* what these invariants are, based on the actual control flow execution of the model.

If we run tracing with real tensor values, we can determine what the true value of `x.size(0)` is, and we can subsequently determine which path the conditional goes down. We can then suggest to the user that they should insert `torch._check(x.size(0) == 20)` after the allocation of x, which will give you a program that will trace without real sizes.

Some details to deal with:

* Hints may be inconsistent with 0/1 specialization. When we have `x = y.nonzero()`, we assume `x.size(0) not in [0, 1]`. User must therefore carefully craft their data-set to always have intermediate values that are >= 2. I can't think of a better way to deal with this: if the real data results in a 1-size tensor, we must not generate guards forcing it to be equal to one, because those guards won't generalize. But if we ignore the real data, then it's not entirely clear how to evaluate the truth value of guards. `x.size(0) == 0` is certainly false, but what about `x.size(0) * 512 == 0` (should be false)? How about `x.size(0) + y.size(0) == y.size(0)`? Perhaps if you can do https://github.com/pytorch/pytorch/pull/105535 that would work, but this is implemented by allocating an unbacked tensor, which is precisely what we're trying to get rid of.
* We don't necessarily need to evaluate real values for everything. For example, in KJT use case, you probably only need the offsets/lengths tensors to be fully materialized to work everything else out. Additionally, if you have some logic which isn't doing conditionals on unbacked symints, you can probably get away with using a normal unbacked symint. The real value is just a get out of jail free card to figure out which way the conditional should resolve.

### Versions

main

**Comment by aakhundov** [comment]
I'm wondering if we'd have something like this instead:

```python
# z is a model input
x = y.nonzero()
if x.size(0) != z.size(0):
     assert False
do_something()
```

where `z.size(0)` would be a backed SymInt (e.g., `s0`), would the above still work?

**Comment by ezyang** [comment]
Yeah, this will work, because z.size(0) will be s0 and so we'll end up with a "guard" mentioning s0, not 20. However, it will be a little annoying to give a good UX here because we need to work out how to spell s0 in a user readable manner (the source will be something like `L["z"].size(0)`

**Comment by lezcano** [comment]
We already do this if the code
```python
if x.size(0) != 20:
     assert False
```
is written as a `torch._check`. Is this issue still relevant?

**Comment by ezyang** [comment]
Yes, this is for conditional branches that we don't know which direction they go

**Comment by ezyang** [comment]
This is maybe a little less pressing as all of our heavy unbacked symint tracing workstreams have successfully managed graph capture

**Comment by pianpwk** [comment]
[issue scrubbing] checking we want to keep this open? draft export exists but nothing concrete for compile
